p2p/base_reactor.go

```diff
@@ -41,25 +46,151 @@ type Reactor interface {
+// unmarshalling the messages and calling Receive on the reactor.
+func WithProcessor(processor func(<-chan UnprocessedEnvelope) error) ReactorOptions {
+	return func(br *BaseReactor) {
+		br.processor = processor
+	}
+}
+
+// WithIncomingQueueSize sets the size of the incoming message queue for a
+// reactor.
+func WithIncomingQueueSize(size int) ReactorOptions {
+	return func(br *BaseReactor) {
+		br.incoming = make(chan UnprocessedEnvelope, size)
 	}
 }
 
 func (br *BaseReactor) SetSwitch(sw *Switch) {
 	br.Switch = sw
 }
+
+// QueueUnprocessedEnvelope is called by the switch when an unprocessed
+// envelope is received. Unprocessed envelopes are immediately buffered in a
+// queue to avoid blocking. The size of the queue can be changed by passing
+// options to the base reactor.
+func (br *BaseReactor) QueueUnprocessedEnvelope(e UnprocessedEnvelope) {
+	br.incoming <- e
+}
+
+// DefaultProcessor unmarshals the message and calls Receive on the reactor.
+// This preservers the sender's original order for all messages.
+func DefaultProcessor(impl Reactor) func(<-chan UnprocessedEnvelope) error {
+	return func(incoming <-chan UnprocessedEnvelope) error {
+		implChannels := impl.GetChannels()
+
+		chIDs := make(map[byte]proto.Message, len(implChannels))
+		for _, chDesc := range implChannels {
+			chIDs[chDesc.ID] = chDesc.MessageType
+		}
+
+		var (
+			err error
+			msg proto.Message
+		)
+
+		for ue := range incoming {
+			mt := chIDs[ue.ChannelID]
+
+			if mt == nil {
+				return fmt.Errorf("no message type registered for channel %d", ue.ChannelID)
+			}
+
+			msg = proto.Clone(mt)
+
+			err = proto.Unmarshal(ue.Message, msg)
+			if err != nil {
+				return fmt.Errorf("unmarshaling message: %v into type: %s", err, reflect.TypeOf(mt))
+			}
+
+			if w, ok := msg.(types.Unwrapper); ok {
+				msg, err = w.Unwrap()
+				if err != nil {
+					return fmt.Errorf("unwrapping message: %v", err)
+				}
+			}
+
+			ue.Src.Metrics().PeerReceiveBytesTotal.
+				With("peer_id", string(ue.Src.ID()), "chID", ue.Src.ChIDToMetricLabel(ue.ChannelID)).
+				Add(float64(len(ue.Message)))
+
+			ue.Src.Metrics().MessageReceiveBytesTotal.
+				With("message_type", ue.Src.ValueToMetricLabel(msg)).
+				Add(float64(len(ue.Message)))
+
+			impl.Receive(Envelope{
+				ChannelID: ue.ChannelID,
+				Src:       ue.Src,
+				Message:   msg,
+			})
+		}
+		return err
+	}
+}
+
+// ParallelProcessor creates a processor that runs multiple goroutines to
+// process incoming messages concurrently. It uses the default processor to
+// unmarshal the messages and call Receive on the reactor. This breaks the
+// guarantee that messages passed to this reactor are processed in the order
+// that the sender sent them.
+func ParallelProcessor(impl Reactor, threads int) func(<-chan UnprocessedEnvelope) error {
+	return func(incoming <-chan UnprocessedEnvelope) error {
+		for i := 0; i < threads; i++ {
+			go func() {
+				DefaultProcessor(impl)(incoming)
+			}()
+		}
+		return nil
+	}
+}
```

Comment:
tbh, I'm not sure if there's a ton of usecases for this atm. The thinking was that we could use this for reactors that don't require processing msgs in the original send order. Initially, I thought the mempool reactor would be a good candidate, and it might!, however it could result in more issues for users that are sending multiple txs within the same block for chains that use the default replay protection in the cosmos-sdk. 

there is a different flavor of this that could be where each channel is processed in order, but the others aren't. We could do this, but we might be better off just refactoring the entire p2p stack to use QUIC streams, since that builds on primitives designed for that.